sign.go

```diff
@@ -0,0 +1,82 @@
+package floodsub
+
+import (
+	"fmt"
+
+	pb "github.com/libp2p/go-floodsub/pb"
+
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+func verifyMessageSignature(m *pb.Message) error {
+	var pubk crypto.PubKey
+
+	pid, err := peer.IDFromBytes(m.From)
+	if err != nil {
+		return err
+	}
+
+	if m.Key == nil {
+		// no attached key, it must be extractable from the source ID
+		pubk, err = pid.ExtractPublicKey()
+		if err != nil {
+			return err
+		}
+	} else {
+		pubk, err = crypto.UnmarshalPublicKey(m.Key)
+		if err != nil {
+			return err
+		}
+
+		// verify that the source ID matches the attached key
+		if !pid.MatchesPublicKey(pubk) {
+			return fmt.Errorf("bad signing key; source ID %s doesn't match key", pid)
+		}
+	}
+
+	xm := pb.Message{
```

Comment:
I wonder if we can do something where we enforce that the protobuf puts the signature at the end, and simply slice the raw bytes to the right point instead of remarshaling

Reply:
that would be very nice indeed -- i don't like the remarshal either.

Reply:
we could also have a pre-allocated buffer where we `WriteMsg` into instead of marshaling.

Reply:
1. Let's not rely on slicing protobufs, if at all possible.
2. Please read: https://github.com/ipfs/notes/issues/249. We need to include *something* that says that the signature is for pubsub.

Regardless, we need to make sure we can add new fields and have them covered by the signature by default.

Reply:
Hrm, not convinced it's absolutely necessary as this is transient data, but sure we can add a prefix.

Reply:
Added a `libp2p-pubsub:` prefix.